Big Refactor: introduce unit tests, upgrade dependencies, improve code locality

official-sounding opened the pull request:
NB: I have run some initial integration/user testing on this, but I also acknowledge I've touched a ton of code and I am only ~90% certain everything is working, especially around the skater editing UI/applying to an existing statsbook flow.  I think this will need some more testing from you @AdamSmasherDerby before merging.

**Goals**
1. Continue move toward more modularization of code, making it easier to find where something is being done
2. Add Unit Tests to improve confidence that changes are doing what is expected
3. Reduce the number of globals, make the event handling code easier to follow
4. Just do file management and UI updates in index.js, have actual "business logic" happen elsewhere.

**Highlights of changes**
1. Restructured code flow to move a lot of the processing code out into separate modules.  this means that...
2. Unit Tests (using jasmine) have been introduced for pretty much everything in xlsxWriter except for score and lineups.
3. Added a wrapper around xlsx-populate, to improve test-ability.
4. Tried to promisify and linearize the code flow within index to make things much easier to follow.
5. Upgraded dependencies, moved to electron 8.0, updated lodash
6. Removed .eslintrc from .gitignore, added an .eslintrc that encompasses (what I think are) the style rules for the project, mostly so I won't litter semicolons all over the place.
7. Removed references to jQuery within index

AdamSmasherDerby commented:
I will get to this when I can, but things are really busy at the moment.  I promise I'm not ignoring it.

AdamSmasherDerby commented:
OK, looks like you duplicated my original error that was fixed in the last release - box trips aren't being handled properly for the pivot and jammer in jams where there is a star pass.

Specifically, in (your version) xlsxWriter.js, it appears that writeBoxTrips is called at line 266 for the no star pass case, but in the star pass branch (starting at 268), only the POST star pass symbols are written for the jammer.

I don't think they're being sorted out correctly for the pivot, either, but that'll probably fall out of fixing the jammer ones.

For the main branch, this was fixed by added a snippet to write pre-star pass box symbols BEFORE the if (!starPass) conditional.  In other words, ALWAYS write the pre-star pass symbols, then check on the post star pass symbols during the conditional.

I like having the writeBoxTrips routine generally, but it needs to be careful about the special case of pivots and jammers switching columns during a star pass.

official-sounding committed:
start breaking out writer into separate class

official-sounding committed:
halfway done on penalties

official-sounding committed:
add eslintrc to repo

official-sounding committed:
colophon + clock data. start on score + lineups

official-sounding committed:
score + lineups work

official-sounding committed:
more score + lineups

official-sounding committed:
cleanup code in score-and-lineups

official-sounding committed:
integrate xlsxWriter with index

official-sounding committed:
start adding skaterManager code

official-sounding committed:
continue skaterManager

official-sounding committed:
update skater lists to be array instead of hash

official-sounding committed:
add test infrastructure

official-sounding committed:
add validation methods to skaterManager

official-sounding committed:
updates to make xlsxWriter work

official-sounding committed:
make xlsxwriter unit testable, and start writing unit tests

put wrapper around workbook to give it an easier to test interface
write a test jig to that interface, and start adding tests
(game info and skaters so far)

official-sounding committed:
penalty page test cases

official-sounding committed:
penalties tests

official-sounding committed:
update dependencies

official-sounding committed:
finish out penalty test scenarios

official-sounding committed:
fix a few workflow items around skaterManager

official-sounding committed:
remove jquery from index.js

official-sounding committed:
upgrade to electron 8.0

official-sounding committed:
update box trip handling to be more explicit

official-sounding committed:
add stubs for scoring test cases

official-sounding commented:
Finally got a chance to look at this again.  I think the concerns with box trips are effectively handled in 93003fa0.  The solution ended up being a bit more complicated than I would have liked, but I think box trip tracking is ultimately going to be more complicated than I would like 😄. At the very least I think the new code is more explicit.

Also took the opportunity to rebase onto the top of current master, and add some additional notes for test cases that were squirreled away on another computer.
